bot/utils/redis_cache.py

```diff
@@ -120,41 +130,49 @@ async def get(self, key: ValidRedisType, default: Optional[ValidRedisType] = Non
         if value is None:
             return default
         else:
-            value = self._from_typestring(value.decode("utf-8"))
+            value = self._from_typestring(value)
             return value
 
     async def delete(self, key: ValidRedisType) -> None:
         """Delete an item from the Redis cache."""
-        # await self._redis.hdel(self._namespace, key)
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        return await self._redis.hdel(self._namespace, key)
 
     async def contains(self, key: ValidRedisType) -> bool:
         """Check if a key exists in the Redis cache."""
-        # return await self._redis.hexists(self._namespace, key)
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        return await self._redis.hexists(self._namespace, key)
 
     async def items(self) -> AsyncIterator:
         """Iterate all the items in the Redis cache."""
-        # data = await redis.hgetall(self.get_with_namespace(key))
-        # for item in data:
-        #     yield item
+        await self._validate_cache()
+        data = await self._redis.hgetall(self._namespace)  # Get all the keys
+        for key, value in self._dict_from_typestring(data).items():
+            yield key, value
 
     async def length(self) -> int:
         """Return the number of items in the Redis cache."""
-        # return await self._redis.hlen(self._namespace)
+        await self._validate_cache()
+        return await self._redis.hlen(self._namespace)
 
     async def to_dict(self) -> Dict:
         """Convert to dict and return."""
-        # return dict(self.items())
+        return {key: value async for key, value in self.items()}
 
     async def clear(self) -> None:
         """Deletes the entire hash from the Redis cache."""
-        # await self._redis.delete(self._namespace)
+        await self._validate_cache()
+        await self._redis.delete(self._namespace)
 
     async def pop(self, key: ValidRedisType, default: Optional[ValidRedisType] = None) -> ValidRedisType:
         """Get the item, remove it from the cache, and provide a default if not found."""
-        # value = await self.get(key, default)
-        # await self.delete(key)
-        # return value
+        value = await self.get(key, default)
+        await self.delete(key)
```

Comment:
Is deleting a key that does not exist "safe"? The `default` implies that trying to pop an non-existing key is valid.

Seems like it's supported, from the [HDEL docs](https://redis.io/commands/hdel):

> Removes the specified fields from the hash stored at key. Specified fields that do not exist within this hash are ignored. If key does not exist, it is treated as an empty hash and this command returns 0.

Reply:
Yeah, I saw the same docs, so I figure it's fine.